### ISSUE TITLE
On android, the event handler stops working when you click on a text field

### ISSUE THREAD

**Issue by Neizvestnyj** [kivy/kivy#7374]
* Python: 3.8.5
* OS: Android
* Kivy: 2.0.0
* Kivy installation method: pip

When you click on the input field, the event handler that is declared in the `init` method stops working, but if you open the control center after that, as shown in the video below, the handler starts working until you click on the text field again. This bug only occurs on android.
_The bug is present on all architectures_

```
from kivy.app import App
from kivy.uix.textinput import TextInput
from kivy.core.window import Window


class MainApp(App):
    def __init__(self, **kwargs):
        super(MainApp, self).__init__(**kwargs)

        Window.bind(on_keyboard=self.events)

    def build(self):
        return TextInput(text="Hello, World")

    def events(self, instance, keyboard, keycode, text, modifiers):
        if keyboard in (1001, 27):
            print('back')
        return True


MainApp().run()
```

2-nd example
```
from kivy.app import App
from kivy.uix.button import Button
from kivy.lang.builder import Builder
from kivy.core.window import Window

KV = '''
Screen:
	ScreenManager:
		id: screen_manager
		Screen:
			name: 'home'
			BoxLayout:
				padding: 100
				Button:
					text: 'Open text input screen'
					on_release: app.open_input_screen()
		Screen:
			name: 'input'
			BoxLayout:
				orientation: 'vertical'
				padding: 50
				spacing: 20
				Button:
					id: btn
					text: 'Text input'
				TextInput:
					id: text_input
					text: 'Text'
	
'''


class TestApp(App):

    def build(self):
        Window.bind(on_keyboard=self.events)
        return Builder.load_string(KV)

    def open_input_screen(self):
        self.root.ids.screen_manager.current = 'input'

    def events(self, instance, keyboard, keycode, text, modifiers):
        if keyboard in (1001, 27):  # back
            if self.root.ids.text_input.text == 'Text':
                self.root.ids.text_input.text = ')'
                return True
            if self.root.ids.screen_manager.current == 'input':
                self.root.ids.screen_manager.current = 'home'
            else:
                quit()
            return True
        return False


TestApp().run()
```

https://user-images.githubusercontent.com/40869738/107146635-8c7e6980-695a-11eb-86ea-304d7e3b036f.mp4

https://user-images.githubusercontent.com/40869738/107230355-09c7de00-6a30-11eb-89f3-490541a56b96.mp4

[buildozer.spec](https://github.com/kivy/kivy/files/5944100/buildozer.txt)

**Comment by matham** [comment]
I believe this is expected because when the text widget is handling the keyboard, the keyboard events don't go to other handlers.

**Comment by Neizvestnyj** [comment]
But as you can see from the video, I close the input, take a closer look, even if I switch to another screen or do something else, events will not work

**Comment by Neizvestnyj** [comment]
> I believe this is expected because when the text widget is handling the keyboard, the keyboard events don't go to other handlers.
But as you can see from the video, I close the input, take a closer look, even if I switch to another screen or do something else, events will not work. tshirtman said this is a bug and asked to open the issue

**Comment by matham** [comment]
Ok, I misread it to be the opposite of what you described. However, the answer is basically the same. Your `events` method returns True indicating that it handled the event, hence it breaks the text input handler.

**Comment by tshirtman** [comment]
Hm, isn't it from our text input management on android, that wouldn't release control, and prevent over key handling, until the app itself loses focus?

**Comment by Neizvestnyj** [comment]
I understand that to work correctly, you need to always `return False`

**Comment by Neizvestnyj** [comment]
But this is not true if I have a lot of windows and need to go between them

**Comment by Neizvestnyj** [comment]
In my app something like that does not work:
    ```
def events(self, instance, keyboard, keycode, text, modifiers):
        if keyboard in (1001, 27):  # back
            if ....':
                if ...:
                    ...
                else:
                    ...
                return True
            if ...:
                return True
            ...
            return True

        return False
```

**Comment by matham** [comment]
Hmm, are you saying that when you switch away from kivy window and interact with other apps, kivy gets the keys typed to other apps until you switch back?

If so, that is indeed a bug. But I could not understand from the video what you're trying to show. Can you please expand a little more on what exactly the issue is?

**Comment by tshirtman** [comment]
That's not what i understand, maybe i'm wrong but:
- app handle events like "back"
- text input is selected, handles text, all good
- keyboard is closed, textinput is unfocused
- app doesn't handle events like "back" anymore
- switching focus away from the whole app, then back again
- app handles events like "back" correctly.

am i correct?

**Comment by Neizvestnyj** [comment]
> Hmm, are you saying that when you switch away from kivy window and interact with other apps, kivy gets the keys typed to other apps until you switch back?
> 
> If so, that is indeed a bug. But I could not understand from the video what you're trying to show. Can you please expand a little more on what exactly the issue is?

Look, I click on the input field, the keyboard appears, pressing the back button, the keyboard closes, but pressing back again, nothing happens, and for example another window should open

**Comment by Neizvestnyj** [comment]
> That's not what i understand, maybe i'm wrong but:
> 
> * app handle events like "back"
> * text input is selected, handles text, all good
> * keyboard is closed, textinput is unfocused
> * app doesn't handle events like "back" anymore
> * switching focus away from the whole app, then back again
> * app handles events like "back" correctly.
> 
> am i correct?

Yes

**Comment by matham** [comment]
Does the problem go away if you don't bind the `events` callback or always return false from it?

**Comment by Neizvestnyj** [comment]
> Does the problem go away if you don't bind the `events` callback or always return false from it?

If you do not bind an event, then nothing will happen, if always returning False this will help, but this is impossible, because otherwise in some of the conditions where you need to switch the screen or execute the function, the application will simply close

**Comment by matham** [comment]
>If you do not bind an event, then nothing will happen, if always returning False this will help, but this is impossible, because otherwise in some of the conditions where you need to switch the screen or execute the function, the application will simply close

This is why I don't think it's a bug (but of course I could be wrong). You're returning `True` for something when you should return `False` if you want the app to be able to handle `back`. I.e. maybe it's ok if you handle the back key the first time and return True, but then you should return False so the framework can handle it.

I don't see a bug here. Do you @tshirtman ?

**Comment by Neizvestnyj** [comment]
> > If you do not bind an event, then nothing will happen, if always returning False this will help, but this is impossible, because otherwise in some of the conditions where you need to switch the screen or execute the function, the application will simply close
> 
> This is why I don't think it's a bug (but of course I could be wrong). You're returning `True` for something when you should return `False` if you want the app to be able to handle `back`. I.e. maybe it's ok if you handle the back key the first time and return True, but then you should return False so the framework can handle it.
> 
> I don't see a bug here. Do you @tshirtman ?

It is strange that the error occurs only on android

**Comment by tshirtman** [comment]
> > If you do not bind an event, then nothing will happen, if always returning False this will help, but this is impossible, because otherwise in some of the conditions where you need to switch the screen or execute the function, the application will simply close
> 
> This is why I don't think it's a bug (but of course I could be wrong). You're returning `True` for something when you should return `False` if you want the app to be able to handle `back`. I.e. maybe it's ok if you handle the back key the first time and return True, but then you should return False so the framework can handle it.
> 
> I don't see a bug here. Do you @tshirtman ?

The fact that it seems dependent on the keyboard having been used for textinput makes me think it's a bug, i would suspect our hidden native TextInput (created by the sdl view) not getting unfocused, and so still getting keyboard events until the app loses focus, or our keyboard handling code in kivy, but i didn't dig.

**Comment by Neizvestnyj** [comment]
See, here's an example from my working app where I'm not returning what?
```
    def events(self, instance, keyboard, keycode, text, modifiers):
        if keyboard in (1001, 27):  # back
            tab = self.root.ids.bottom_nav.previous_tab.name  # get previous tab screen
            if self.root.ids.screen_manager.current == 'scr login' or self.root.ids.screen_manager.current == 'login home':
                if self.root.ids.login.current is 'forgot_password':
                    self.root.ids.login.current = 'main'
                else:
                    self.open_home_scr()
                return True
            if self.img_for_mode_screen is not None:  # if open image manager
                self.cancel_image()
                return True
            if self.root.ids.screen_manager.current == 'scr modes':
                self.modes_list()
                return True
            elif self.root.ids.screen_manager.current == 'scr support':
                self.open_home_scr()
                return True
            elif self.root.ids.screen_manager.current == 'scr lang':
                self.open_lang_screen()  # open home screen
                return True
            if tab == 'screen home':  # bottom navigation tabs
                if self.back_click_counter == 0:
                    Clock.schedule_once(self.zero_counter, 2.0)
                    self.back_click_counter += 1
                else:
                    self.close_app()
            elif tab == 'screen bg':
                self.root.ids.bottom_nav.switch_tab('screen home')
            elif tab == 'screen settings':
                self.root.ids.bottom_nav.switch_tab('screen home')
            return True

        return False
```

**Comment by tshirtman** [comment]
I have no idea what your message mean, that's probably not going to help solving the issue, please make sure we don't have to guess things when you describe your issue.

**Comment by Neizvestnyj** [comment]
above you described the problem correctly

**Comment by matham** [comment]
I'm not very familiar with the android side, but you're using the 1001 key code and I imagine it's an internal key that you shouldn't use. I can't seem to find this number being used anywhere so I'm not sure where it's coming from or why it's there. Or perhaps 27 (escape) is the issue.

But I'm not sure I can help further given that it's not clear to me how kivy handles the back button or escape internally on android so it'll require someone to investigate. So I'll just re-open.

**Comment by Neizvestnyj** [comment]
> I'm not very familiar with the android side, but you're using the 1001 key code and I imagine it's an internal key that you shouldn't use. I can't seem to find this number being used anywhere so I'm not sure where it's coming from or why it's there. Or perhaps 27 (escape) is the issue.
> 
> But I'm not sure I can help further given that it's not clear to me how kivy handles the back button or escape internally on android so it'll require someone to investigate. So I'll just re-open.

As I said above, everything works on Window, Linux, but with android the problem, the codes 27, 1001 themselves are esc on all platforms

**Comment by Neizvestnyj** [comment]
@tshirtman, @matham So, I found out that this occurs because of `p4a. branch = develop`. As soon as I uncommented this line and rebuilt the app, events stopped working

**Comment by Neizvestnyj** [comment]
I also added the above example for tests

**Comment by Neizvestnyj** [comment]
This solution does not help https://github.com/SomberNight/python-for-android/commit/f2a33b88a5f0ee76dce6c256365065ba20d62325, the solution is not accurate

**Comment by Neizvestnyj** [comment]
Link to the user's video with the same error https://youtu.be/GN_YwHQ6Y0I

**Comment by Neizvestnyj** [comment]
So, I checked again and found out, it's not in the branches, I tried `p4a.branch = release-2020.06.02, develop, master` but in all this bug and it's not in _buildozer_, so I have no idea what the problem is, I looked at the old releases of the application, the bug was also there, I used the standard branch there and _buildozer 1.2.0_ just now updated it with **pip**

**Comment by Neizvestnyj** [comment]
I also tried `kivy 1.11.1`, the same bug, but as I said above, it's an old bug, so maybe it's something else

**Comment by Neizvestnyj** [comment]
Also, maybe this is important, I use this to make the scipy library work https://github.com/mzakharo/android-gfortran/releases/tag/r19c

**Comment by Neizvestnyj** [comment]
I found the problem. I created a clean VM and installed `buildozer 1.2.0. dev0`, after which I made a standard _init_ and built the application, everything worked, after I changed the branch to _develop_ and uncommented p4a. bootstrap = sdl2, re-built the application, and here it is a bug

**Comment by matham** [comment]
If develop is the issue, can you try it with kivy master?

**Comment by Neizvestnyj** [comment]
@matham How I can do it?

**Comment by Neizvestnyj** [comment]
I tried to comment out `# p4a. bootstrap = sdl2` and rebuild, it didn't help

**Comment by matham** [comment]
In the requirements you have to write something like `kivy=master`. I'm not sure the exact syntax.

**Comment by Neizvestnyj** [comment]
> In the requirements you have to write something like `kivy=master`. I'm not sure the exact syntax.

`kivy==master`, Already rebuilding the package (sdl uncommented)

**Comment by Neizvestnyj** [comment]
I deleted the folder _.buildozer_, installed in the _spec_ file `requirements = python3,kivy== master`, `p4a. branch = develop`, `p4a. bootstrap = sdl2`, and built, the bug is the same

**Comment by Neizvestnyj** [comment]
The most interesting thing is that if after all the manipulations you try to re-build the application with the standard values, deleting the _buildozer_ folder before that, then the bug will be

**Comment by Neizvestnyj** [comment]
I also tried deleting the folder _.buildozer_ to the directory above the application (ie the main one) and rebuild everything with the following arguments `requirements = python3, kivy== master`, `android. api = 30`, `android. arch = armeabi-v7a`, `p4a. branch = develop`,` p4a. bootstrap = sdl2`, the bug remains. To avoid the bug, you need to set the standard arguments and delete the folders `.buildozer` (there are 2 of them). But this is a very bad decision, I am waiting for your comments

**Comment by matham** [comment]
I'm afraid you may need to investigate to see how kivy/py4a handles the back button and how ti gets to the text input (if at all). I don't fully understand why the back button is treated as keyboard input, but I suspect https://github.com/kivy/kivy/blob/master/kivy/support.py#L86 may have something to do with it. But I'm not familiar with how our android implementation works so I can't really be of much help.

**Comment by Neizvestnyj** [comment]
Therefore, I created a issue to find out how to make everything work again

**Comment by Neizvestnyj** [comment]
Another thing I learned, only the back button is not processed, the keyboard works fine

**Comment by Neizvestnyj** [comment]
I tested the builds with most branches on absolutely clean virtual machines and deleting all the files associated with the build, the bug remains, as it seems to me it is possible to solve this problem for a millisecond after closing the keyboard to take the application out of focus, as I said above it helps, but I do not know how to do it and this is a bad decision

**Comment by kengoon** [comment]
I have encountered this issue and it only occurs in android version 10, and maybe 9

**Comment by Neizvestnyj** [comment]
> I have encountered this issue and it only occurs in `android version 10`, and maybe `9` other version works perfectly.
> 
> in my own case, I did not include the 1001 key, but I still got that issue

I also think that there is a bug in the latest version of android 11

**Comment by HentaiLover1337** [comment]
I also have this bug in master branch, api 30, and as I remember, it is not new bug

**Comment by ChihayaK** [comment]
I also encountered this bug when using the text field on API 30(Android 11). However it is working correctly on API 24(Android 7.0).

**Comment by Neizvestnyj** [comment]
@ChihayaK api 24 is deprecated, target development is done using api 29

**Comment by Neizvestnyj** [comment]
Fixed in `kivy==master` (2.2.0dev0), realted issue - https://github.com/kivy/python-for-android/issues/2423
